README
=================================================================
            libdmtx - Open Source Data Matrix Software
=================================================================

               libdmtx README file (all platforms)

This summary of the libdmtx package applies generally to all
platforms. For instructions regarding your specific platform,
also see the README.xxx file in this directory that matches your
system (e.g., README.linux, README.osx, etc...).


1. Introduction
-----------------------------------------------------------------

libdmtx is a software library that enables programs to read and
write Data Matrix barcodes of the modern ECC200 variety. The
library runs natively on several platforms and can be accessed by
multiple languages using the libdmtx language wrappers. The
utility programs dmtxread and dmtxwrite also provide a command
line interface for libdmtx, and serve as a good reference for
developers writing their own libdmtx-enabled programs.

This package (libdmtx) contains only the core library, and is
distributed under a Simplified BSD license with an alternate
waiver option. See the LICENSE file in the main project directory
for full terms of use and distribution.

The non-library components related to libdmtx are available as
separate downloads, and are distributed under a different license
(typically LGPLv2). Please contact support@dragonflylogic.com if
you require clarification on licensing. It's not complicated, but
it's important to us that all license terms are respected (not
just ours).


2. Project Components
-----------------------------------------------------------------

The libdmtx project serves a diverse audience and contains many
components -- some of which may not be useful to you. Components
fall into one of four categories:

  Description        Package        Audience
  -----------------  -------------  ----------------------
  Core library       libdmtx        libdmtx programs
  Test programs      libdmtx        libdmtx developers
  Utility programs   dmtx-utils     Shell and command line
  Language Wrappers  dmtx-wrappers  Non-C/C++ developers


3. Installation
-----------------------------------------------------------------

libdmtx uses GNU Autotools so installation should be familiar to
free software veterans. If your platform cannot easily run the
Autotools scripts, refer to the appropriate platform-specific
README.xxx located in this directory for alternate instructions.

In theory the following 3 steps would build and install libdmtx
on your system:

  $ ./configure
  $ make
  $ sudo make install

However, you may need to install additional software or make
other changes for these steps to work properly. The details below
will help to address errors and/or customize beyond the defaults.

Problems with "configure" step
----------------------------------------
If you obtained libdmtx from Git you may have received an error
like "./configure: No such file or directory". Run this command
before trying again:

  $ ./autogen.sh

The autogen.sh command requires autoconf, automake, libtool, and
pkgconfig to be installed on your system.

The configure script also offers many options for customizing the
build process, described in detail by running:

  $ ./configure --help

Problems with "make" step
----------------------------------------
Errors encountered during the "make" step are often a result of
missing software dependencies. Install any missing software
mentioned in the error message(s) and try again.

Problems with "sudo make install" step
----------------------------------------
If the 'sudo' command is not configured on your system, you can
alternatively yell "Yeeehaww!" as you log in as root and run it
like this:

  # make install

And finally...
----------------------------------------
If you want to verify that everything is working properly you can
optionally build the test programs:

  $ make check

This command will not perform any tests, but will build the
programs that contain test logic: multi_test, rotate_test,
simple_test, and unit_test.

Note: multi_test and rotate_test contain extra dependencies due
to their graphical nature, and are not terribly useful unless you
need to test the library's internals.


5. Contact
-----------------------------------------------------------------

Project website:       www.libdmtx.org
Documentation wiki:    libdmtx.wikidot.com
SourceForge.net page:  www.sourceforge.net/projects/libdmtx
OhLoh.net page:        www.ohloh.net/projects/libdmtx
Open mailing list:     libdmtx-open_discussion@lists.sourceforge.net
Professional support:  www.dragonflylogic.com/products


6. This Document
-----------------------------------------------------------------

This document is derived from the wiki page located at:

  http://libdmtx.wikidot.com/general-instructions

If you find an error or have additional helpful information,
please edit the wiki directly with your updates.


NEWS
26-May-2011: New official download page available
17-Mar-2011: Core library relicensed to Simplified BSD with waiver option
03-Mar-2011: libdmtx code now in SourceForge GIT repositories
29-Nov-2010: libdmtx binaries now available as official Fedora package
18-Nov-2009: Mikko Koppanen released improved PHP bindings for libdmtx
05-Nov-2009: "Onion" released a Gstreamer element using libdmtx
04-Sep-2009: New release: libdmtx-0.7.2 (Release Notes)
06-Aug-2009: Simon Wood released pyDataMatrixScanner at SourceForge
03-Aug-2009: Evan Nemerson added Vala support with a new wrapper
31-Jul-2009: libdmtx Documentation wiki moved to libdmtx.wikidot.com
19-May-2009: libdmtx project discussed in Linux Journal (March 2009)
12-Mar-2009: Python wrapper supports multi-barcode scans (thanks Simon!)
02-Mar-2009: New release: libdmtx-0.7.0 (Release Notes)
31-Jan-2009: Experimental libdmtx Windows binaries now available
23-Nov-2008: New release: libdmtx-0.6.0 (Release Notes)
08-Oct-2008: Olivier Guilyardi added support for all major image formats! (SVN)
04-Sep-2008: New release: libdmtx-0.5.2 (README)
20-Aug-2008: Significant performance gains available in SVN (latest tarball)
01-Jul-2008: New release: libdmtx-0.5.1 (README)
29-Jun-2008: libdmtx.org now hosted on a super efficient fit-PC
03-Jun-2008: New iPhone reader by Stefan Hafeneger (no jailbreak required)
31-May-2008: New libdmtx mailing lists replace forums
30-May-2008: GeoWeb does it again: CameraDMTX for Nokia N60
27-May-2008: Et tu, Cognex?
22-May-2008: Congratulations Cognex! Good guys 2, patent trolls 0
16-Apr-2008: New documentation wiki. Everyone can contribute!
13-Apr-2008: 5th release: libdmtx-0.5.0 (README)
22-Feb-2008: New phone video, this time on Symbian S60
15-Jan-2008: Unofficial .debs now available (thanks pcitizen!)
21-Dec-2007: Check out Martin's iDMTX video!
07-Dec-2007: 4th release: libdmtx-0.4.0 (README)
02-Dec-2007: Good news: Project is ACTIVE - new release soon
16-Aug-2007: Added new screenshots to show progress during freeze
30-Nov-2006: Bad news: Downloads have been halted (explanation)
15-Oct-2006: 3rd release: libdmtx-0.3.0 (README)
13-Sep-2006: Unstable version now available from CVS (instructions)
07-Sep-2006: Now taking feature requests for next release
11-Jun-2006: 2nd release: libdmtx-0.2.0 (view README file before using)
22-Apr-2006: 1st release: libdmtx-0.1.0 (view README file before using)
12-Mar-2006: Project home page changed hosting
24-Jan-2006: ISO/IEC 16022:2000 specification arrived in the mail
25-Nov-2005: Project home page libdmtx.sourceforge.net created
22-Nov-2005: SourceForge project created


dmtxfec.c
/*
libdmtx - Data Matrix Encoding/Decoding Library

Copyright (C) 2002-2004 Phil Karn, KA9Q
Copyright (C) 2008, 2009 Mike Laughton

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

----------------------------------------------------------------------
This file is derived from various portions of the fec library
written by Phil Karn available at http://www.ka9q.net. It has
been modified to include only the specific cases used by Data
Matrix barcodes, and to integrate with the rest of libdmtx.
----------------------------------------------------------------------

Contact: mike@dragonflylogic.com
*/

/* $Id$ */

/**
 * @file dmtxfec.c
 * @brief Forward Error Correction
 */

#include <stdlib.h>
#include <string.h>

typedef unsigned char data_t;

#define DMTX_RS_MM          8 /* Bits per symbol */
#define DMTX_RS_NN        255 /* Symbols per RS block */
#define DMTX_RS_GFPOLY  0x12d /* Field generator polynomial coefficients */
#define DMTX_RS_MAX_NROOTS 68 /* Maximum value of nroots for Data Matrix */

#undef NULL
#define NULL ((void *)0)

/* Reed-Solomon codec control block
 *
 * rs->nroots  - the number of roots in the RS code generator polynomial,
 *               which is the same as the number of parity symbols in a block.
 *               Integer variable or literal.
 * rs->pad     - the number of pad symbols in a block. Integer variable or literal.
 * rs->genpoly - an array of rs->nroots+1 elements containing the generator polynomial in index form
 */
struct rs {
   data_t genpoly[DMTX_RS_MAX_NROOTS+1]; /* Generator polynomial */
   int nroots;                           /* Number of generator roots = number of parity symbols */
   int pad;                              /* Padding bytes in shortened block */
};

/* General purpose RS codec, 8-bit symbols */
static void encode_rs_char(struct rs *rs, unsigned char *data, unsigned char *parity);
static int decode_rs_char(struct rs *rs, unsigned char *data, int *eras_pos, int no_eras, int max_fixes);
static struct rs *init_rs_char(int nroots, int pad);
static void free_rs_char(struct rs **rs);

/* log(0) = -inf */
static unsigned char indexOf[] =
   { 255,   0,   1, 240,   2, 225, 241,  53,   3,  38, 226, 133, 242,  43,  54, 210,
       4, 195,  39, 114, 227, 106, 134,  28, 243, 140,  44,  23,  55, 118, 211, 234,
       5, 219, 196,  96,  40, 222, 115, 103, 228,  78, 107, 125, 135,   8,  29, 162,
     244, 186, 141, 180,  45,  99,  24,  49,  56,  13, 119, 153, 212, 199, 235,  91,
       6,  76, 220, 217, 197,  11,  97, 184,  41,  36, 223, 253, 116, 138, 104, 193,
     229,  86,  79, 171, 108, 165, 126, 145, 136,  34,   9,  74,  30,  32, 163,  84,
     245, 173, 187, 204, 142,  81, 181, 190,  46,  88, 100, 159,  25, 231,  50, 207,
      57, 147,  14,  67, 120, 128, 154, 248, 213, 167, 200,  63, 236, 110,  92, 176,
       7, 161,  77, 124, 221, 102, 218,  95, 198,  90,  12, 152,  98,  48, 185, 179,
      42, 209,  37, 132, 224,  52, 254, 239, 117, 233, 139,  22, 105,  27, 194, 113,
     230, 206,  87, 158,  80, 189, 172, 203, 109, 175, 166,  62, 127, 247, 146,  66,
     137, 192,  35, 252,  10, 183,  75, 216,  31,  83,  33,  73, 164, 144,  85, 170,
     246,  65, 174,  61, 188, 202, 205, 157, 143, 169,  82,  72, 182, 215, 191, 251,
      47, 178,  89, 151, 101,  94, 160, 123,  26, 112, 232,  21,  51, 238, 208, 131,
      58,  69, 148,  18,  15,  16,  68,  17, 121, 149, 129,  19, 155,  59, 249,  70,
     214, 250, 168,  71, 201, 156,  64,  60, 237, 130, 111,  20,  93, 122, 177, 150 };

/* alpha**-inf = 0 */
static unsigned char alphaTo[] =
   {   1,   2,   4,   8,  16,  32,  64, 128,  45,  90, 180,  69, 138,  57, 114, 228,
     229, 231, 227, 235, 251, 219, 155,  27,  54, 108, 216, 157,  23,  46,  92, 184,
      93, 186,  89, 178,  73, 146,   9,  18,  36,  72, 144,  13,  26,  52, 104, 208,
     141,  55, 110, 220, 149,   7,  14,  28,  56, 112, 224, 237, 247, 195, 171, 123,
     246, 193, 175, 115, 230, 225, 239, 243, 203, 187,  91, 182,  65, 130,  41,  82,
     164, 101, 202, 185,  95, 190,  81, 162, 105, 210, 137,  63, 126, 252, 213, 135,
      35,  70, 140,  53, 106, 212, 133,  39,  78, 156,  21,  42,  84, 168, 125, 250,
     217, 159,  19,  38,  76, 152,  29,  58, 116, 232, 253, 215, 131,  43,  86, 172,
     117, 234, 249, 223, 147,  11,  22,  44,  88, 176,  77, 154,  25,  50, 100, 200,
     189,  87, 174, 113, 226, 233, 255, 211, 139,  59, 118, 236, 245, 199, 163, 107,
     214, 129,  47,  94, 188,  85, 170, 121, 242, 201, 191,  83, 166,  97, 194, 169,
     127, 254, 209, 143,  51, 102, 204, 181,  71, 142,  49,  98, 196, 165, 103, 206,
     177,  79, 158,  17,  34,  68, 136,  61, 122, 244, 197, 167,  99, 198, 161, 111,
     222, 145,  15,  30,  60, 120, 240, 205, 183,  67, 134,  33,  66, 132,  37,  74,
     148,   5,  10,  20,  40,  80, 160, 109, 218, 153,  31,  62, 124, 248, 221, 151,
       3,   6,  12,  24,  48,  96, 192, 173, 119, 238, 241, 207, 179,  75, 150,   0 };

/**
 * @brief  Calculate x mod NN (255)
 * @param  x
 * @return Modulus
 */
static int
modnn(int x)
{
   while (x >= DMTX_RS_NN) {
      x -= DMTX_RS_NN;
      x = (x >> DMTX_RS_MM) + (x & DMTX_RS_NN);
   }

   return x;
}

/**
 * @brief  Free memory used for Reed Solomon struct
 * @param  p
 * @return rs
 */
static void
free_rs_char(struct rs **p)
{
   if(*p != NULL)
      free(*p);

   *p = NULL;
}

/**
 * @brief  Initialize a Reed-Solomon codec
 * @param  nroots RS code generator polynomial degree (number of roots)
 * @param  pad Padding bytes at front of shortened block
 * @return Pointer to rs struct
 */
static struct rs *
init_rs_char(int nroots, int pad)
{
   struct rs *rs;
   int i, j, root;

   assert(nroots <= DMTX_RS_MAX_NROOTS);

   /* Can't have more roots than symbol values */
   if(nroots < 0 || nroots > DMTX_RS_NN)
      return NULL;

   /* Too much padding */
   if(pad < 0 || pad >= (DMTX_RS_NN - nroots))
      return NULL;

   rs = (struct rs *)calloc(1, sizeof(struct rs));
   if(rs == NULL)
      return NULL;

   /* Form RS code generator polynomial from its roots */
   rs->pad = pad;
   rs->nroots = nroots;

   rs->genpoly[0] = 1;
   for(i = 0, root = 1; i < nroots; i++, root++) {
      rs->genpoly[i+1] = 1;

      /* Multiply rs->genpoly[] by @**(root + x) */
      for(j = i; j > 0; j--) {
         rs->genpoly[j] = (rs->genpoly[j] == 0) ? rs->genpoly[j-1] :
               rs->genpoly[j-1] ^ alphaTo[modnn(indexOf[rs->genpoly[j]] + root)];
      }

      /* rs->genpoly[0] can never be zero */
      rs->genpoly[0] = alphaTo[modnn(indexOf[rs->genpoly[0]] + root)];
   }

   /* convert rs->genpoly[] to index form for quicker encoding */
   for(i = 0; i <= nroots; i++)
      rs->genpoly[i] = indexOf[rs->genpoly[i]];

   return rs;
}

/**
 * @brief  Encode array of values with Reed-Solomon
 * @param  rs Reed-Solomon codec control block
 * @param  data Array of data_t to be encoded
 * @param  parity Array of data_t to be written with parity symbols
 * @return void
 */
static void
encode_rs_char(struct rs *rs, data_t *data, data_t *parity)
{
   int i, j;
   data_t feedback;

   memset(parity, 0, rs->nroots * sizeof(data_t));

   for(i = 0; i < DMTX_RS_NN - rs->nroots - rs->pad; i++) {
      feedback = indexOf[data[i] ^ parity[0]];
      if(feedback != DMTX_RS_NN) { /* feedback term is non-zero */
         for(j = 1; j < rs->nroots; j++)
            parity[j] ^= alphaTo[modnn(feedback + rs->genpoly[rs->nroots - j])];
      }

      /* Shift */
      memmove(&parity[0], &parity[1], sizeof(data_t) * (rs->nroots-1));
      if(feedback != DMTX_RS_NN)
         parity[rs->nroots-1] = alphaTo[modnn(feedback + rs->genpoly[0])];
      else
         parity[rs->nroots-1] = 0;
   }
}

/**
 * @brief  Decode and correct array of values with Reed-Solomon
 * @param  rs Reed-Solomon codec control block
 * @param  data Array of data and parity symbols to be corrected in place
 * @param  eras_pos
 * @param  no_eras
 * @param  max_fixes
 * @return count
 */
static int
decode_rs_char(struct rs *rs, data_t *data, int *eras_pos, int no_eras, int max_fixes)
{
   int deg_lambda, el, deg_omega;
   int i, j, r, k;
   data_t u, q, tmp, num1, num2, den, discr_r;
   int syn_error, count;

   /* Err+Eras Locator poly and syndrome poly */
   data_t lambda[DMTX_RS_MAX_NROOTS+1];
   data_t s[DMTX_RS_MAX_NROOTS];
   data_t b[DMTX_RS_MAX_NROOTS+1];
   data_t t[DMTX_RS_MAX_NROOTS+1];
   data_t omega[DMTX_RS_MAX_NROOTS+1];
   data_t root[DMTX_RS_MAX_NROOTS];
   data_t reg[DMTX_RS_MAX_NROOTS+1];
   data_t loc[DMTX_RS_MAX_NROOTS];

   /* Form the syndromes; i.e., evaluate data(x) at roots of g(x) */
   for(i = 0; i < rs->nroots; i++)
      s[i] = data[0];

   for(j = 1; j < DMTX_RS_NN - rs->pad; j++) {
      for(i = 0; i< rs->nroots; i++) {
         s[i] = (s[i] == 0) ? data[j] : data[j] ^ alphaTo[modnn(indexOf[s[i]] + (1+i))];
      }
   }

   /* Convert syndromes to index form, checking for nonzero condition */
   syn_error = 0;
   for(i = 0; i < rs->nroots; i++) {
      syn_error |= s[i];
      s[i] = indexOf[s[i]];
   }

   /* If syndrome is zero, data[] is a codeword and there are no errors to
      correct. So return data[] unmodified */
   if(syn_error == 0) {
      count = 0;
      goto finish;
   }
   else if(max_fixes == 0) {
      count = -1;
      goto finish;
   }

   memset(&lambda[1], 0, rs->nroots * sizeof(lambda[0]));
   lambda[0] = 1;

   if(no_eras > 0) {
      /* Init lambda to be the erasure locator polynomial */
      lambda[1] = alphaTo[modnn(DMTX_RS_NN - 1 - eras_pos[0])];
      for(i = 1; i < no_eras; i++) {
         u = modnn(DMTX_RS_NN - 1 - eras_pos[i]);
         for(j = i+1; j > 0; j--) {
            tmp = indexOf[lambda[j - 1]];
            if(tmp != DMTX_RS_NN)
               lambda[j] ^= alphaTo[modnn(u + tmp)];
         }
      }
   }
   for(i = 0; i < rs->nroots + 1; i++)
      b[i] = indexOf[lambda[i]];

   /* Begin Berlekamp-Massey algorithm to determine error+erasure
      locator polynomial */
   r = no_eras;
   el = no_eras;
   while(++r <= rs->nroots) { /* r is the step number */
      /* Compute discrepancy at the r-th step in poly-form */
      discr_r = 0;
      for(i = 0; i < r; i++) {
         if((lambda[i] != 0) && (s[r-i-1] != DMTX_RS_NN)) {
            discr_r ^= alphaTo[modnn(indexOf[lambda[i]] + s[r-i-1])];
         }
      }
      discr_r = indexOf[discr_r]; /* Index form */
      if(discr_r == DMTX_RS_NN) {
         /* 2 lines below: B(x) <-- x*B(x) */
         memmove(&b[1], b, rs->nroots * sizeof(b[0]));
         b[0] = DMTX_RS_NN;
      }
      else {
         /* 7 lines below: T(x) <-- lambda(x) - discr_r*x*b(x) */
         t[0] = lambda[0];
         for(i = 0 ; i < rs->nroots; i++) {
            t[i+1] = (b[i] != DMTX_RS_NN) ? lambda[i+1] ^ alphaTo[modnn(discr_r + b[i])] : lambda[i+1];
         }
         if(2 * el <= r + no_eras - 1) {
            el = r + no_eras - el;
            /* 2 lines below: B(x) <-- inv(discr_r) * lambda(x) */
            for(i = 0; i <= rs->nroots; i++)
               b[i] = (lambda[i] == 0) ? DMTX_RS_NN : modnn(indexOf[lambda[i]] - discr_r + DMTX_RS_NN);
         }
         else {
            /* 2 lines below: B(x) <-- x*B(x) */
            memmove(&b[1], b, rs->nroots * sizeof(b[0]));
            b[0] = DMTX_RS_NN;
         }
         memcpy(lambda, t, (rs->nroots+1) * sizeof(t[0]));
      }
   }

   /* Convert lambda to index form and compute deg(lambda(x)) */
   deg_lambda = 0;
   for(i = 0; i < rs->nroots + 1; i++) {
      lambda[i] = indexOf[lambda[i]];
      if(lambda[i] != DMTX_RS_NN)
         deg_lambda = i;
   }

   /* Find roots of the error+erasure locator polynomial by Chien search */
   memcpy(&reg[1], &lambda[1], rs->nroots * sizeof(reg[0]));
   count = 0; /* Number of roots of lambda(x) */
   for(i = 1,k=0; i <= DMTX_RS_NN; i++,k = modnn(k+1)) {
      q = 1; /* lambda[0] is always 0 */
      for(j = deg_lambda; j > 0; j--) {
         if(reg[j] != DMTX_RS_NN) {
            reg[j] = modnn(reg[j] + j);
            q ^= alphaTo[reg[j]];
         }
      }

      if(q != 0)
         continue; /* Not a root */

      /* Store root (index-form) and error location number */
      root[count] = i;
      loc[count] = k;

      /* If we've already found max possible roots, abort the search to save time */
      if(++count == deg_lambda)
         break;
   }

   if(deg_lambda != count) {
      /* deg(lambda) unequal to number of roots => uncorrectable error detected */
      count = -1;
      goto finish;
   }

   /* Compute err+eras evaluator poly omega(x) = s(x)*lambda(x)
      (modulo x**rs->nroots). in index form. Also find deg(omega). */
   deg_omega = deg_lambda-1;
   for(i = 0; i <= deg_omega;i++) {
      tmp = 0;
      for(j=i;j >= 0; j--) {
         if((s[i - j] != DMTX_RS_NN) && (lambda[j] != DMTX_RS_NN))
            tmp ^= alphaTo[modnn(s[i - j] + lambda[j])];
      }
      omega[i] = indexOf[tmp];
   }

   /* Compute error values in poly-form. num1 = omega(inv(X(l))), num2 =
      inv(X(l))**0 and den = lambda_pr(inv(X(l))) all in poly-form */
   for(j = count - 1; j >= 0; j--) {
      num1 = 0;
      for(i = deg_omega; i >= 0; i--) {
         if(omega[i] != DMTX_RS_NN)
            num1 ^= alphaTo[modnn(omega[i] + i * root[j])];
      }
      num2 = alphaTo[modnn(DMTX_RS_NN)];
      den = 0;

      /* lambda[i+1] for i even is the formal derivative lambda_pr of lambda[i] */
      for(i = min(deg_lambda, rs->nroots - 1) & ~1; i >= 0; i -=2) {
         if(lambda[i+1] != DMTX_RS_NN)
            den ^= alphaTo[modnn(lambda[i+1] + i * root[j])];
      }

      /* Apply error to data */
      if(num1 != 0 && loc[j] >= rs->pad) {
         data[loc[j] - rs->pad] ^= alphaTo[modnn(indexOf[num1] +
               indexOf[num2] + DMTX_RS_NN - indexOf[den])];
      }
   }

finish:
   if(eras_pos != NULL) {
      for(i = 0; i < count; i++)
         eras_pos[i] = loc[i];
   }

   return count;
}
